Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt

```diff
@@ -30,32 +34,49 @@ class BlockChain(var blockList: ArrayList<Block> = arrayListOf()) {
 
 
     override fun toString(): String {
-        return blockList.joinToString (separator = "\n"){ it.print() }
+        return blockList.joinToString(separator = "\n") { it.print() }
     }
 
-    fun checkBlock(blockInvestigating: Block, previousBlock: Block): Boolean{
-        if (previousBlock.hash != blockInvestigating.previousBlockHash){
+    private fun checkBlock(blockInvestigating: Block, previousBlock: Block): Boolean {
+        if (previousBlock.hash != blockInvestigating.previousBlockHash) {
+            return false
+        }
+        try {
+            blockInvestigating.transactions.forEach {
+                interpreter.runContract(it.data)
+            }
+        } catch (ex: Exception) {
             return false
         }
         return true
     }
 
-    fun isValid(): Boolean{
-        for (i in 1 until (blockList.size -1)){
-            val isValid = checkBlock(blockList[i], blockList[i-1])
-            if(!isValid){
+    fun isValid(): Boolean {
+        for (i in 1 until (blockList.size - 1)) {
+            val isValid = checkBlock(blockList[i], blockList[i - 1])
+            if (!isValid) {
                 return false
             }
         }
         return true
     }
 
-    fun replaceChain(newBlockChain: BlockChain): Boolean{
-        if (newBlockChain.isValid()){
+    fun replaceChain(newBlockChain: BlockChain): Boolean {
+        if (newBlockChain.isValid()) {
             this.blockList = newBlockChain.blockList
             return true
         }
         return false
     }
 
+    fun getBalanceFromAddress(address: String): Int {
+        return interpreter.accountList.filter {
+            it.address == address
+        }[0].weight
+    }
+
+    fun listKnownAddresses(): Array<Account> {
+        return interpreter.accountList.toArray() as Array<Account>
```

Comment:
This has poor type safety. I couldn't get it to agree with me, I'll hack at it again, but if you can come up with anything to get rid of the warning feel free to propose a change.

Reply:
Pushed an edit for this.